### PR TITLE
cmd: move tikv-server and tikv-ctl to cmd crate

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? (mandatory)

Move tikv-server and tikv-ctl to the new `cmd` crate.

After this PR, `tikv-server` can depends on crates which depends on the `tikv` crate. Eg. we can integrate importer to tikv-server. Also this PR should improve compile-time.

## What are the type of changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Manually.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect `tidb-ansible` update? (mandatory)

No.

### CODE REVIEW

**Comment by siddontang** [comment]
how about the compiling time now?

**Comment by siddontang** [comment]
btw, I think it is strange to move these to the components. IMO, the components should be the base crates which TiKV depends on. 

So for the binary, I think we can move to another directory besides components and src, may be cmd? 

/cc @BusyJay

File: components/cmd/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+name = "cmd"
+version = "0.0.1"
+license = "Apache-2.0"
+edition = "2018"
+publish = false
+
+[features]
+default = ["tikv/default"]
+tcmalloc = ["tikv/tcmalloc"]
+jemalloc = ["tikv/jemalloc", "tikv/jemalloc"]
```
Comment (5kbpers) [inline components/cmd/Cargo.toml]: Duplicate definition.

**Comment by overvenus** [comment]
> IMO, the components should be the base crates which TiKV depends on.

Some crates under components are actully depends on tikv, eg, test_xxx.

> how about the compiling time now?

Well, it's almost the same (slightly improved).
Before
```bash
   Compiling tikv v4.0.0-alpha (/home/tikv)
    Finished dev [unoptimized + debuginfo] target(s) in 6m 13s
➜  tikv git:(master) touch src/binutil/mod.rs
➜  tikv git:(master) cargo build --bin tikv-server
    Compiling tikv v4.0.0-alpha (/home/tikv)
    Finished dev [unoptimized + debuginfo] target(s) in 53.71s
```

Now
```bash
   Compiling tikv v4.0.0-alpha (/home/tikv)
   Compiling cmd v0.0.1 (/home/tikv/components/cmd)
    Finished dev [unoptimized + debuginfo] target(s) in 5m 55s
➜  tikv git:(cmd) touch components/cmd/src/lib.rs
➜  tikv git:(cmd) cargo build -p cmd --bin tikv-server
   Compiling cmd v0.0.1 (/home/tikv/components/cmd)
    Finished dev [unoptimized + debuginfo] target(s) in 51.39s
```

**Comment by siddontang** [comment]
> Some crates under components are actully depends on tikv, eg, test_xxx.

If this, we need to do the refactor, not follow the bad style. we should make the hierarchy more clear, no circle dependent. 

IMO, the dependency may be components <- TiKV <- (bin, test)

**Comment by overvenus** [comment]
> If this, we need to do the refactor, not follow the bad style. we should make the hierarchy more clear, no circle dependent.
>
> IMO, the dependency may be components <- TiKV <- (bin, test)

By putting cmd under components does not mean circle dependent. Rust crate dependency management is not based on dir hierarchy.

**Comment by siddontang** [comment]
> By putting cmd under components does not mean circle dependent. Rust crate dependency management is not based on dir hierarchy.

that is why our hierarchy is ugly and messy. If we can separate them forcibly and clearly, why not do it?

File: cmd/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+name = "cmd"
+
+[[bin]]
+name = "tikv-server"
+
+[[bin]]
+name = "tikv-ctl"
+
+[dependencies]
+tikv = { path = "../" }
```
Comment (BusyJay) [inline cmd/Cargo.toml]: Should disable the default features.

**Comment by overvenus** [comment]
PTAL, thanks!

File: cmd/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+futures = "0.1"
+serde = "1.0"
+serde_json = "1.0"
+fail = "0.3"
+grpcio = { version = "0.5.0-alpha.3", features = [ "openssl-vendored" ] }
+raft = "0.6.0-alpha"
+hex = "0.3"
+vlog = "0.1.4"
+kvproto = { git = "https://github.com/pingcap/kvproto.git" }
+
```
Comment (BusyJay) [inline cmd/Cargo.toml]: Why extra blank line?

Comment (overvenus) [reply cmd/Cargo.toml]: Copied from tikv's Cargo.toml.

**Pull request merged**